modules/core/kivent_core/systems/animation.pyx

```diff
@@ -0,0 +1,215 @@
+        '''
+        Function to initialise an AnimationComponent
+
+        Optional Args:
+
+            name (str): name of the animation which is registered in animation_manager
+            loop (bool): whether to loop this animation or not
+        '''
+        model_manager = self.gameworld.model_manager
+        animation_manager = self.gameworld.animation_manager
+
+        cdef MemoryZone memory_zone = self.imz_components.memory_zone
+        cdef AnimationStruct* component = <AnimationStruct*>(
+            memory_zone.get_pointer(component_index))
+        cdef FrameList frame_list = animation_manager.animations[args['name']]
+        component.entity_id = entity_id
+        component.frames = <void*>frame_list
+        component.current_frame_index = 0
+        component.current_duration = 0
+        component.loop = args['loop']
+
+        return self.entity_components.add_entity(entity_id, zone)
+
+    def clear_component(self, unsigned int component_index):
+        cdef MemoryZone memory_zone = self.imz_components.memory_zone
+        cdef AnimationStruct* pointer = <AnimationStruct*>(
+            memory_zone.get_pointer(component_index))
+        pointer.entity_id = -1
+        pointer.frames = NULL
+        pointer.current_frame_index = <unsigned int>-1
+        pointer.current_duration = 0
+
+
+    def remove_component(self, unsigned int component_index):
+        cdef AnimationComponent component = self.components[component_index]
+        self.entity_components.remove_entity(component.entity_id)
+        super(AnimationSystem, self).remove_component(component_index)
+
+
+    def update(self, dt):
+        gameworld = self.gameworld
+        cdef void** component_data = <void**>(
+            self.entity_components.memory_block.data)
+        cdef unsigned int component_count = self.entity_components.count
+        cdef unsigned int count = self.entity_components.memory_block.count
+        cdef unsigned int i, real_index
+        cdef FrameList frame_list
+        cdef Frame frame
+        cdef AnimationStruct* anim_comp
+        cdef RenderStruct* render_comp
+        cdef FrameStruct* frame_data, next_frame
+        cdef unsigned int current_index
+        cdef unsigned int groupkey
+        cdef VertexModel new_model
+        cdef Renderer renderer
+        cdef float u0, u1, v0, v1
+        cdef list uv_list
+
+
+        for i in range(count):
+            real_index = i*component_count
+            if component_data[real_index] == NULL:
+                continue
+            anim_comp = <AnimationStruct*>component_data[real_index]
+            render_comp = <RenderStruct*>component_data[real_index+1]
+
+            frame_list = <FrameList>anim_comp.frames
+            current_index = anim_comp.current_frame_index
+            frame = frame_list[current_index]
+            frame_data = <FrameStruct*>frame.frame_pointer
+
+            if current_index != <unsigned int>-1 and anim_comp.dirty:
```

Comment:
we should switch this logic so that it occurs after the below logic that sets a comp to dirty, that way it occurs immediately after it happening instead of waiting a frame.